Query string parameters to RedirectToRoute, Query-* property group in RouteLink

Rumbeer opened the pull request:
Added option to specify query string parameters in `RedirectToRoute` method.

Added UI test for this feature.

Issue https://github.com/riganti/dotvvm/issues/405

exyi commented on src/DotVVM.Framework/Hosting/DotvvmRequestContextExtensions.cs:
```diff
@@ -83,14 +85,37 @@ public static void RedirectToUrl(this IDotvvmRequestContext context, string url,
         /// Returns the redirect response and interrupts the execution of current request.
         /// </summary>
-        public static void RedirectToRoute(this IDotvvmRequestContext context, string routeName, object newRouteValues = null, bool replaceInHistory = false, bool allowSpaRedirect = true, string urlSuffix = null)
+        public static void RedirectToRoute(this IDotvvmRequestContext context, string routeName, object newRouteValues = null, bool replaceInHistory = false, bool allowSpaRedirect = true, string urlSuffix = null, object queryStringParameters = null)
```
I'd rather use a shorter name because this parameter is going to be used as a named one, what about `query`? I think it's quite clear what it means the context of redirection.

exyi commented on src/DotVVM.Framework/Hosting/DotvvmRequestContextExtensions.cs:
```diff
@@ -83,14 +85,37 @@ public static void RedirectToUrl(this IDotvvmRequestContext context, string url,
+                foreach (KeyValuePair<string, string> item in (IEnumerable<KeyValuePair<string, string>>)queryStringParameters)
+                {
+                    resultSuffix = resultSuffix + (urlSuffix.LastIndexOf('?') < 0 ? "?" : "&") + Regex.Escape(item.Key) +
+                                "=" + Regex.Escape(item.Value);
```
`Regex.Escape` IMO does not make any sense for escaping urls, did you wanted to use `Uri.EscapeDataString` instead?

exyi commented on src/DotVVM.Framework/Hosting/DotvvmRequestContextExtensions.cs:
```diff
@@ -83,14 +85,37 @@ public static void RedirectToUrl(this IDotvvmRequestContext context, string url,
+        {
+            var hashIndex = urlSuffix.IndexOf('#');
+            var resultSuffix = hashIndex < 0 ? urlSuffix : urlSuffix.Substring(0, hashIndex);
+            if (queryStringParameters is IEnumerable<KeyValuePair<string, string>>)
```
I hope you could use C# 7 pattern matching feature, it would make the code a bit more repetitive in this case ;) https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#pattern-matching

Rumbeer commented:
Added `Query-*` property group to `RouteLink` that use the same method for building the suffix when the control is rendered on server.

Added UI test for this feature.

Issue https://github.com/riganti/dotvvm/issues/404

exyi commented on src/DotVVM.Framework/Resources/Scripts/DotVVM.ts:
```diff
@@ -755,6 +755,28 @@ class DotVVM {
+            if (query.hasOwnProperty(property)) {
+                if (!property) continue;
+                var queryParamValue = ko.unwrap(query[property]);
+                if (!queryParamValue) continue;
```
This will break when the query would contain number `0` or boolean `false`, you should probably use `queryParamValue != null` condition. It will `continue;` for `null` and `undefined` and will include the parameter for `0`, empty string and `false`
